src/libfetchers/git.cc

```diff
@@ -21,6 +21,15 @@ namespace nix::fetchers {
 // old version of git, which will ignore unrecognized `-c` options.
 const std::string gitInitialBranch = "__nix_dummy_branch";
 
+static std::string getGitDir()
+{
+    auto gitDir = getEnv("GIT_DIR");
+    if (!gitDir) {
+        return ".git";
+    }
+    return *gitDir;
+}
+
 static std::string readHead(const Path & path)
 {
     return chomp(runProgram("git", true, { "-C", path, "--git-dir", ".git", "rev-parse", "--abbrev-ref", "HEAD" }));
```

Comment:
Is there a reason why this shouldn’t have been changed as well?

```suggestion
    return chomp(runProgram("git", true, { "-C", path, "--git-dir", getGitDir(), "rev-parse", "--abbrev-ref", "HEAD" }));
```